File: knowledge_inferer/inference_engine.py
import pandas as pd
import json
import os
from typing import Dict, Any, Optional, List
from datetime import datetime

from .phase1_data_standardization import DataStandardizer
from .phase2_indicator_calculation import IndicatorCalculator
from .phase3_risk_evaluation import RiskEvaluator
from .phase4_composite_inference import CompositeInferer


class InferenceEngine:
    """
    Main Inference Engine
    Điều phối toàn bộ quy trình suy diễn 4 pha
    """
    
    def __init__(self, config_dir: str):
        """
        Args:
            config_dir: Thư mục chứa các file cấu hình JSON
        """
        self.config_dir = config_dir
        
        schema_path = os.path.join(config_dir, 'schema_mapping.json')
        calculation_path = os.path.join(config_dir, 'calculation_rules.json')
        indicator_path = os.path.join(config_dir, 'indicator_rules.json')
        composite_path = os.path.join(config_dir, 'composite_rules.json')
        
        self.phase1 = DataStandardizer(schema_path)
        self.phase2 = IndicatorCalculator(calculation_path)
        self.phase3 = RiskEvaluator(indicator_path)
        self.phase4 = CompositeInferer(composite_path)
        
        self.known_facts = None
        self.solution_log = []
    
    def run_full_pipeline(self, raw_data_path: str) -> pd.DataFrame:
        """
        Chạy toàn bộ pipeline suy diễn
        
        Args:
            raw_data_path: Đường dẫn đến file dữ liệu thô
            
        Returns:
            DataFrame chứa kết quả đầy đủ
        """
        print("\n" + "="*60)
        print("FINANCIAL RISK INFERENCE ENGINE")
        print("="*60)
        
        self.known_facts = self.phase1.standardize(raw_data_path)
        
        self.known_facts = self.phase2.calculate_batch(self.known_facts)
        
        self.known_facts = self.phase3.evaluate_batch(self.known_facts)
        
        self.known_facts = self.phase4.infer_batch(self.known_facts)
        
        print("="*60)
        print("INFERENCE COMPLETED SUCCESSFULLY")
        print("="*60 + "\n")
        
        return self.known_facts
    
    def infer_single_company(self, 
                            raw_data_path: str, 
                            ticker: str, 
                            year: int, 
                            quarter: int) -> Dict[str, Any]:
        """
        Suy diễn cho một công ty cụ thể
        
        Args:
            raw_data_path: Đường dẫn đến file dữ liệu thô
            ticker: Mã chứng khoán
            year: Năm báo cáo
            quarter: Quý báo cáo
            
        Returns:
            Dict chứa kết quả suy diễn đầy đủ
        """
        print(f"\n{'='*60}")
        print(f"Analyzing: {ticker} - Q{quarter}/{year}")
        print(f"{'='*60}\n")
        
        standardized_data = self.phase1.standardize(raw_data_path)
        
        company_data = self.phase1.get_company_data(standardized_data, ticker, year, quarter)
        
        if company_data is None:
            return {
                'error': 'Company data not found',
                'ticker': ticker,
                'year': year,
                'quarter': quarter
            }
        
        print("\n=== Phase 2: Calculating Indicators ===")
        indicators = self.phase2.calculate_all_indicators(company_data)
        for ind, val in indicators.items():
            print(f"{ind}: {val:.4f}" if not pd.isna(val) else f"{ind}: N/A")
        
        print("\n=== Phase 3: Evaluating Risks ===")
        evaluations = self.phase3.evaluate_all_indicators(indicators)
        for ind, eval_result in evaluations.items():
            print(f"{ind}: {eval_result['risk_level']} (points: {eval_result['risk_point']})")
        
        risk_summary = self.phase3.get_risk_summary(evaluations)
        category_summary = self.phase3.get_category_summary(evaluations)
        
        print("\n=== Phase 4: Composite Risk Inference ===")
        composite_risks = self.phase4.infer_composite_risks(evaluations)
        print(f"Detected {len(composite_risks)} composite risks:")
        for risk in composite_risks:
            print(f"- [{risk['rule_id']}] {risk['risk_type']} ({risk['severity']})")
        
        final_score = self.phase4.calculate_final_risk_score(evaluations, composite_risks)
        explanation = self.phase4.generate_explanation(evaluations, composite_risks)
        
        result = {
            'company_info': {
                'ticker': ticker,
                'year': year,
                'quarter': quarter
            },
            'indicators': indicators,
            'evaluations': evaluations,
            'risk_summary': risk_summary,
            'category_summary': category_summary,
            'composite_risks': composite_risks,
            'final_score': final_score,
            'explanation': explanation,
            'timestamp': datetime.now().isoformat()
        }
        
        print(f"\n{'='*60}")
        print(f"FINAL RISK ASSESSMENT")
        print(f"{'='*60}")
        print(f"Risk Score: {final_score['risk_score']:.2%}")
        print(f"Risk Level: {final_score['risk_level']}")
        print(f"Total Risk Points: {final_score['total_risk_points']}/{final_score['max_total_points']}")
        print(f"Composite Risks: {final_score['composite_risk_count']}")
        print(f"{'='*60}\n")
        
        return result
    
    def export_results(self, output_path: str, format: str = 'csv'):
        """
        Xuất kết quả ra file
        
        Args:
            output_path: Đường dẫn file đầu ra
            format: Định dạng file ('csv', 'excel', 'json')
        """
        if self.known_facts is None:
            print("No data to export. Run inference first.")
            return
        
        if format == 'csv':
            export_df = self.known_facts.copy()
            
            export_df['composite_risks_json'] = export_df['composite_risks'].apply(
                lambda x: json.dumps(x, ensure_ascii=False)
            )
            export_df = export_df.drop(columns=['composite_risks'])
            
            export_df.to_csv(output_path, index=False, encoding='utf-8-sig')
            print(f"✓ Results exported to {output_path}")
        
        elif format == 'excel':
            export_df = self.known_facts.copy()
            export_df['composite_risks_json'] = export_df['composite_risks'].apply(
                lambda x: json.dumps(x, ensure_ascii=False)
            )
            export_df = export_df.drop(columns=['composite_risks'])
            
            export_df.to_excel(output_path, index=False, engine='openpyxl')
            print(f"✓ Results exported to {output_path}")
        
        elif format == 'json':
            export_data = self.known_facts.to_dict(orient='records')
            
            with open(output_path, 'w', encoding='utf-8') as f:
                json.dump(export_data, f, ensure_ascii=False, indent=2, default=str)
            print(f"✓ Results exported to {output_path}")
    
    def get_statistics(self) -> Dict[str, Any]:
        """Lấy thống kê tổng quan"""
        if self.known_facts is None:
            return {}
        
        df = self.known_facts
        
        stats = {
            'total_companies': len(df),
            'total_records': len(df),
            'risk_level_distribution': df['composite_risk_count'].describe().to_dict(),
            'average_composite_risks': df['composite_risk_count'].mean(),
            'max_composite_risks': df['composite_risk_count'].max(),
            'companies_with_high_risk': (df['composite_risk_count'] >= 3).sum()
        }
        
        return stats

File: src/core/InferenceEngine.js
/**
 * Inference Engine - Tích hợp 4 phases
 * Pipeline hoàn chỉnh từ raw data đến kết quả cuối cùng
 */

import { DataStandardizer } from '../phases/Phase1DataStandardization.js';
import { IndicatorCalculator } from '../phases/Phase2IndicatorCalculation.js';
import { RiskEvaluator } from '../phases/Phase3RiskEvaluation.js';
import { CompositeInferer } from '../phases/Phase4CompositeInference.js';
import path from 'path';

export class InferenceEngine {
  constructor(configDir, useForwardChaining = true) {
    this.configDir = configDir;
    this.useForwardChaining = useForwardChaining;

    // Initialize all 4 phases
    this.phase1 = new DataStandardizer(
      path.join(configDir, 'schema_mapping.json')
    );
    this.phase2 = new IndicatorCalculator(
      path.join(configDir, 'indicator_rules.json')
    );
    this.phase3 = new RiskEvaluator(
      path.join(configDir, 'risk_signal_rules.json')
    );
    this.phase4 = new CompositeInferer(
      path.join(configDir, 'risk_label_rules.json'),
      useForwardChaining
    );
  }

  /**
   * Chạy inference cho một công ty
   */
  async inferSingleCompany(rawDataPath, ticker, year, quarter) {
    console.log('\n' + '='.repeat(80));
    console.log(`INFERENCE ENGINE - ${ticker} Q${quarter}/${year}`);
    console.log('='.repeat(80));

    // Phase 1: Data Standardization
    const standardizedData = await this.phase1.standardize(rawDataPath);
    const companyData = this.phase1.getCompanyData(standardizedData, ticker, year, quarter);

    // Phase 2: Indicator Calculation
    const indicators = this.phase2.calculateAll(companyData);

    // Phase 3: Risk Evaluation
    const evaluations = this.phase3.evaluateAll(indicators);

    // Phase 4: Composite Inference
    const phase4Result = this.phase4.infer(evaluations);

    // Kết hợp kết quả từ 4 phases
    const result = {
      company_info: {
        ticker: ticker,
        year: year,
        quarter: quarter
      },
      indicators: indicators,
      evaluations: evaluations,
      composite_risks: phase4Result.composite_risks,
      final_score: phase4Result.final_score,
      inference_stats: phase4Result.inference_stats,
      inference_trace: phase4Result.inference_trace || [],
      backward_chaining: phase4Result.backward_chaining || {}
    };

    this._printSummary(result);
    return result;
  }

  /**
   * Chạy inference cho nhiều công ty (batch)
   */
  async inferBatch(rawDataPath, companies) {
    console.log('\n' + '='.repeat(80));
    console.log('BATCH INFERENCE');
    console.log('='.repeat(80));

    const results = [];

    for (const company of companies) {
      try {
        const result = await this.inferSingleCompany(
          rawDataPath,
          company.ticker,
          company.year,
          company.quarter
        );
        results.push(result);
      } catch (error) {
        console.error(`Error processing ${company.ticker}: ${error.message}`);
      }
    }

    return results;
  }

  /**
   * So sánh nhiều công ty
   */
  async compareCompanies(rawDataPath, tickers, year, quarter) {
    console.log('\n' + '='.repeat(80));
    console.log(`COMPANY COMPARISON - Q${quarter}/${year}`);
    console.log('='.repeat(80));

    const results = {};

    for (const ticker of tickers) {
      try {
        const result = await this.inferSingleCompany(rawDataPath, ticker, year, quarter);
        results[ticker] = result;
      } catch (error) {
        console.error(`Error processing ${ticker}: ${error.message}`);
      }
    }

    // Ranking
    const ranking = Object.entries(results)
      .map(([ticker, result]) => ({
        ticker: ticker,
        risk_score: result.final_score.risk_score,
        risk_level: result.final_score.risk_level
      }))
      .sort((a, b) => a.risk_score - b.risk_score);

    console.log('\n=== RANKING (Best to Worst) ===');
    for (let i = 0; i < ranking.length; i++) {
      const company = ranking[i];
      console.log(
        `${i + 1}. ${company.ticker}: ${(company.risk_score * 100).toFixed(2)}% (${company.risk_level})`
      );
    }

    return {
      results: results,
      ranking: ranking
    };
  }

  /**
   * Giải thích một fact cụ thể (delegate to Phase4)
   */
  explainFact(factId) {
    return this.phase4.explainFact(factId);
  }

  /**
   * Query backward chaining (delegate to Phase4)
   */
  queryBackward(goal, evaluations) {
    return this.phase4.queryBackward(goal, evaluations);
  }

  /**
   * Export inference trace (delegate to Phase4)
   */
  exportInferenceTrace(outputPath) {
    return this.phase4.exportInferenceTrace(outputPath);
  }

  /**
   * Lấy explanation summary (delegate to Phase4)
   */
  getExplanationSummary() {
    return this.phase4.getExplanationSummary();
  }

  /**
   * In tóm tắt kết quả
   */
  _printSummary(result) {
    console.log('\n' + '='.repeat(80));
    console.log('FINAL RESULT');
    console.log('='.repeat(80));

    const finalScore = result.final_score;
    console.log(`\nRisk Score: ${(finalScore.risk_score * 100).toFixed(2)}%`);
    console.log(`Risk Level: ${finalScore.risk_level}`);
    console.log(`Total Points: ${finalScore.total_risk_points}/${finalScore.max_total_points}`);

    if (result.composite_risks && result.composite_risks.length > 0) {
      console.log(`\nComposite Risks Detected: ${result.composite_risks.length}`);
      for (const risk of result.composite_risks) {
        console.log(`  - [${risk.rule_id}] ${risk.risk_type} (${risk.severity})`);
      }
    }

    if (result.inference_stats) {
      console.log(`\nInference Statistics:`);
      console.log(`  - Iterations: ${result.inference_stats.iterations}`);
      console.log(`  - Rules fired: ${result.inference_stats.rules_fired}`);
    }

    console.log('\n' + '='.repeat(80));
  }
}
